### PR TITLE
Fix authentication in Az.LabServices module to work in Azure Automation

### PR DESCRIPTION
**Pull request by petehauge**
Azure Automation now works with Managed Identities (this is the default when creating a new account) instead of the Run As accounts that existed before.  We needed some changes to our module to work in Azure Automation & MSIs correctly!

### CODE REVIEW

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+# Only set these if not running under Azure Automation
+if ("AzureAutomation/" -ne $env:AZUREPS_HOST_ENVIRONMENT) {
+    # We are using strict mode for added safety
+    Set-StrictMode -Version Latest
 
-# We require a relatively new version of Powershell
-#requires -Version 3.0
+    # We require a relatively new version of Powershell
+    #requires -Version 3.0
+}
```
Comment (petehauge) [inline samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:11]: The commandlet and directive seem to work differently in Azure Automation than in regular PowerShell.  The only way I could make it work is excluding these lines if we're running in an Azure Automation environment...

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 [hunk 2]
```diff
@@ -213,32 +216,50 @@ function Get-AzureRmCachedAccessToken() {
     $ErrorActionPreference = 'Stop'
     Set-StrictMode -Off
 
-    $azureRmProfile = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile
-    if (-not $azureRmProfile.Accounts.Count) {
-        Write-Error "Ensure you have logged in before calling this function."
+    if ("AzureAutomation/" -eq $env:AZUREPS_HOST_ENVIRONMENT) {
```
Comment (petehauge) [inline samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:219]: When running in Azure Automation environment, we need to get the token another way (doing a direct Rest API call based on a localhost identity endpoint).  This is based on Automation accounts that use Run As, Service Principal, User-provided MSI or System-provided MSI.

**Review by RogerBestMsft — APPROVED** [review]

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 (hunk 2 above)
Comment (EMaher) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:219]: @petehauge - Code itself for calling the identity endpoint looks good.

However, before we add this, did you test call `Connect-AzAccount -Identity`?  It's the supported cmdlet to connect using an MSI.  See https://docs.microsoft.com/en-us/powershell/module/az.accounts/Connect-AzAccount?view=azps-7.3.2#example-5--connect-using-a-managed-service-identity.   

If running, `Connect-AzAccount -Identity` doesn't work, then code looks good to me.

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 (hunk 1 above)
Comment (EMaher) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:11]: Weird.  Honestly, do we even need this anymore?

**Review by EMaher — APPROVED** [review]

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 (hunk 1 above)
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:11]: I don't think so - I haven't heard of any customers on super old versions of PowerShell any longer...  I'm fine removing it!

File: samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1 (hunk 2 above)
Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:219]: Yes-  I agree that's the way to login - the user would need to call Connect-AzAccount -Identity in their runbook (I tested that this works), but in the library we specifically need the access token which isn't available after calling the commandlet to login with an MSI...  I'll try again using Connect-AzAccount and see if the access token is populated and let you know!
https://docs.microsoft.com/en-us/azure/automation/enable-managed-identity-for-automation

Comment (EMaher) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:219]: Sounds good.  I just remember having a similar issue with the Azure PowerShell functions and that call fixed the issue.  

If that does work, I wonder what the best way to educate people would be.

Comment (petehauge) [reply samples/ClassroomLabs/Modules/Library/Az.LabServices.psm1:219]: @EMaher  - good call on this, I tried with simplified code relying on Connect-AzAccount -Identity and it all worked as expected!  I've simplified the code and will get it checked in shortly after some additional testing.

**Pull request merged**